Check real client IP for access when behind a proxy

renan opened the pull request:
The module ngx_http_access_module allows limiting access to certain
*connecting* client addresses. When using it behind a proxy it would
then use the proxy IP instead of the real client IP, which is ultimately
the desired outcome.

By using the module ngx_http_geo_module we can explicitly specify what
are the proxy IPs thus allowing us to check the real client IP directly.

This implementation also creates unique geo variables to be used by
locations having the same set of whitelisted IPs. This can result in a
much smaller configuration file.

The nginx documentation also recommends the usage of geo variables when
having a lot of rules:
> In case of a lot of rules, the use of the ngx_http_geo_module module variables is preferable.

See:
- http://nginx.org/en/docs/http/ngx_http_access_module.html
- http://nginx.org/en/docs/http/ngx_http_geo_module.html

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation only

## How Has This Been Tested?

Apart from the added test cases, we (Instapro Group) have been successfully running a custom image with those changes in our Kubernetes clusters for a couple of weeks.

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/kubernetes/ingress-nginx/blob/main/CONTRIBUTING.md) guide
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

renan commented:
Hey @longwuyuan!

With `externalTrafficPolicy: Local` you become limited to only accepting traffic in the nodes running the ingress-nginx pods. This in itself is not a problem, I believe it can be a problem when you are scaling up and down constantly, which is my case, and the pods are being recreated in different nodes. You need to have a good health check for that: to ensure requests don't go to nodes without those pods.

In my case, this doesn't really work anyway. This MR targets specific setups running behind an (external) proxy. Having my clusters running behind a proxy, like Cloudflare, the connecting IP would still be the Cloudflare origin servers and not the real client. When doing it, it won't use `$remote_addr` but, with the combination of configuring `enable-real-ip: true` and `proxy-real-ip-cidr` it will trust the header `X-Forwarded-For` (or `CF-Connecting-IP` for Cloudflare) thus being able to get the real client IP to check for permissions.

In my setup it is even more complex because I have: `Internet -> Proxy/Cloudflare -> Load Balancer/AWS NLB -> Node -> Pod`. With those changes I don't think setups using `externalTrafficPolicy: Local` will be affected.

So, in summary. While `externalTrafficPolicy: Local` might inadvertently work on getting the real client IP for some setups. Using `ngx_http_geo_module` would bring a broader range of support.

I hope it is a bit clearer. :)

longwuyuan commented:
Hi @renan,
Thank you for the details. Its certainly more clear now.

Another curiosity, will those 2 modules load for all users of the ingress-nginx-controller, by default ?

renan commented:
/remove-lifecycle stale

rikatz commented:
/ok-to-test
/lgtm
/approve
/hold
@renan thanks for the contribution! Can you please just rebase/wait until https://github.com/kubernetes/ingress-nginx/pull/8795 is merged?

renan commented:
@rikatz Thank you for the follow-up!

I've rebased my branch and incorporated the denylist changes to work like the whitelist.

Let me know what you think of it.

renan commented:
We are not interested in this feature anymore.
